translator/opencensus/oc_to_internal.go

```diff
@@ -0,0 +1,382 @@
+)
+
+func ocToInternal(td consumerdata.TraceData) data.TraceData {
+
+	if td.Node == nil && td.Resource == nil && len(td.Spans) == 0 {
+		return data.TraceData{}
+	}
+
+	resource := ocNodeResourceToInternal(td.Node, td.Resource)
+
+	resourceSpans := data.NewResourceSpans(resource, nil)
+	resourceSpanList := []*data.ResourceSpans{resourceSpans}
+
+	spanCount := len(td.Spans)
+	if spanCount == 0 {
+		return data.NewTraceData(resourceSpanList)
+	}
+
+	// Create slice that holds pointers and another slice that holds structs at once
+	// to avoid individual allocations of structs.
+	spans := data.NewSpanSlice(spanCount)
+	spanPtrs := make([]*data.Span, 0, spanCount)
+
+	for i, ocSpan := range td.Spans {
+		if ocSpan == nil {
+			// Skip nil spans.
+			continue
+		}
+
+		// Point one element in slice of pointers to an element in slice of structs.
+		destSpan := &spans[i]
+
+		ocSpanToInternal(destSpan, ocSpan)
+
+		if ocSpan.Resource != nil {
+			// Add a separate ResourceSpans item just for this span since it
+			// has a different Resource.
+			separateRS := data.NewResourceSpans(
+				ocNodeResourceToInternal(nil, ocSpan.Resource),
+				[]*data.Span{destSpan},
+			)
+			resourceSpanList = append(resourceSpanList, separateRS)
+		} else {
+			// Otherwise add the span to the first ResourceSpans item.
+			spanPtrs = append(spanPtrs, destSpan)
+		}
+	}
+
+	resourceSpans.SetSpans(spanPtrs)
+
+	return data.NewTraceData(resourceSpanList)
+}
+
+func ocSpanToInternal(dest *data.Span, src *octrace.Span) {
+	events, droppedEventCount := ocEventsToInternal(src.TimeEvents)
+	links, droppedLinkCount := ocLinksToInternal(src.Links)
+
+	dest.SetTraceID(data.TraceIDFromBytes(src.TraceId))
+	dest.SetSpanID(data.SpanIDFromBytes(src.SpanId))
+	dest.SetTraceState(ocTraceStateToInternal(src.Tracestate))
+	dest.SetParentSpanID(data.SpanIDFromBytes(src.ParentSpanId))
+	dest.SetName(src.Name.GetValue())
+	dest.SetKind(ocSpanKindToInternal(src.Kind, src.Attributes))
+	dest.SetStartTime(internal.TimestampToUnixnano(src.StartTime))
+	dest.SetEndTime(internal.TimestampToUnixnano(src.EndTime))
+	dest.SetAttributes(ocAttrsToInternal(src.Attributes))
+	dest.SetEvents(events)
+	dest.SetDroppedEventsCount(droppedEventCount)
+	dest.SetLinks(links)
+	dest.SetDroppedLinksCount(droppedLinkCount)
+	dest.SetStatus(ocStatusToInternal(src.Status))
+}
+
+func ocStatusToInternal(ocStatus *octrace.Status) data.SpanStatus {
+	if ocStatus == nil {
+		return data.SpanStatus{}
+	}
+	return data.NewSpanStatus(data.StatusCode(ocStatus.Code), ocStatus.Message)
+}
+
+// Convert tracestate to W3C format. See the https://w3c.github.io/trace-context/
+func ocTraceStateToInternal(ocTracestate *octrace.Span_Tracestate) data.TraceState {
+	if ocTracestate == nil {
+		return ""
+	}
+	var sb strings.Builder
+	for i, entry := range ocTracestate.Entries {
+		if i > 0 {
+			sb.WriteString(",")
+		}
+		sb.WriteString(entry.Key)
+		sb.WriteString("=")
+		sb.WriteString(entry.Value)
+	}
```

Comment:
A small trick to reduce number of allocations, add this as the first operation in the loop:
`sb.Grow(1 + len(entry.Key) + 1 + len(entry.Value));`

Reply:
done